pkg/controller/servicemesh/controlplane/reconciler.go

```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
+			}
+		}
+	}
+	return base
+}
+
+func (r *ControlPlaneReconciler) getSMCPTemplate(name string) (v1.ControlPlaneSpec, error) {
+	if strings.Contains(name, "/") {
+		return v1.ControlPlaneSpec{}, fmt.Errorf("template name contains invalid character '/'")
+	}
+
+	templateContent, err := ioutil.ReadFile(smcpTemplatePath + name)
+	if err != nil {
+		//if we can't read from the user template path, try from the default path
+		//we use two paths because Kubernetes will not auto-update volume mounted
+		//configmaps mounted in directories with pre-existing content
+		defaultTemplateContent, defaultErr := ioutil.ReadFile(smcpDefaultTemplatePath + name)
+		if defaultErr != nil {
+			return v1.ControlPlaneSpec{}, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
+		}
+		templateContent = defaultTemplateContent
+	}
+
+	var template v1.ServiceMeshControlPlane
+	if err = yaml.Unmarshal(templateContent, &template); err != nil {
+		return v1.ControlPlaneSpec{}, fmt.Errorf("failed to parse template %s contents: %s", name, err)
+
+	}
+	return template.Spec, nil
+}
+
+func (r *ControlPlaneReconciler) renderSMCPTemplates(smcp v1.ControlPlaneSpec, visited map[string]struct{}) (v1.ControlPlaneSpec, error) {
+	if smcp.Template == "" {
+		return smcp, nil
+	}
+	r.Log.Info(fmt.Sprintf("processing smcp template %s", smcp.Template))
+
+	if _, ok := visited[smcp.Template]; ok {
+		return smcp, fmt.Errorf("SMCP templates form cyclic dependency. Cannot proceed")
+	}
+
+	template, err := r.getSMCPTemplate(smcp.Template)
+	if err != nil {
+		return smcp, err
+	}
+
+	template, err = r.renderSMCPTemplates(template, visited)
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("error rendering SMCP templates: %s\n", err))
+		return smcp, err
+	}
+
+	visited[smcp.Template] = struct{}{}
+
+	smcp.Istio = mergeValues(smcp.Istio, template.Istio)
+	smcp.ThreeScale = mergeValues(smcp.ThreeScale, template.ThreeScale)
+	return smcp, nil
+}
+
+func (r *ControlPlaneReconciler) updateSMCPWithTemplates(smcpSpec v1.ControlPlaneSpec) (v1.ControlPlaneSpec, error) {
+	r.Log.Info("updating servicemeshcontrolplane with templates")
+	if smcpSpec.Template == "" {
+		smcpSpec.Template = smcpDefaultTemplate
+	}
+
+	spec, err := r.renderSMCPTemplates(smcpSpec, make(map[string]struct{}, 0))
+	r.Log.Info(fmt.Sprintf("finished updating ServiceMeshControlPlane: %+v", spec))
+
+	return spec, err
+}
+
 func (r *ControlPlaneReconciler) renderCharts() error {
 	r.Log.Info("rendering helm charts")
 	allErrors := []error{}
 	var err error
 	var threeScaleRenderings map[string][]manifest.Manifest
 
-	if globalValues, ok := r.Instance.Spec.Istio["global"].(map[string]interface{}); ok {
-		globalValues["operatorNamespace"] = r.OperatorNamespace
-	} else {
-		return fmt.Errorf("Could not set operatorNamespace value, as .Values.global is not a map[string]interface{}: %v", err)
+	if r.Instance.Spec.Istio == nil {
```

Comment:
I think we should be using `r.Instance.Status.LastAppliedConfiguration` here, as mentioned above.  If we're doing that, I don't think any of these other changes are necessary, other than to operate on the status value instead of the spec.

Reply:
The code between here and 341 is a fix for 701. Currently, if an SMCP is created without an Istio section, the operator will crash every time it goes to process the config because there was no check for nil on the Istio field. Also, if a user is using templates, the following a perfectly valid SMCP:

`apiVersion: maistra.io/v1
kind: ServiceMeshControlPlane
metadata:
  name: basic-install
spec:
  template: maistra`

Reply:
Fixed